javaparser-core-testing/src/test/java/com/github/javaparser/ast/validator/CompactClassTest.java

```diff
@@ -0,0 +1,294 @@
+    void compactClassWithField() {
+        String code = "String greeting = \"Hello\";\n" + "void main() { System.out.println(greeting); }";
+        ParseResult<CompilationUnit> result = javaParser.parse(COMPILATION_UNIT, provider(code));
+
+        assertNoProblems(result);
+        assertTrue(result.isSuccessful());
+
+        CompilationUnit cu = result.getResult().get();
+        ClassOrInterfaceDeclaration implicitClass = cu.getClassByName("").get();
+        assertTrue(implicitClass.isCompact());
+
+        assertEquals(1, implicitClass.getFields().size(), "Should have one field");
+        assertEquals(1, implicitClass.getMethods().size(), "Should have one method");
+    }
+
+    @Test
+    void compactClassWithMultipleMethods() {
+        String code = "void main() { helper(); }\n" + "void helper() { System.out.println(\"Helper\"); }";
+        ParseResult<CompilationUnit> result = javaParser.parse(COMPILATION_UNIT, provider(code));
+
+        assertNoProblems(result);
+        assertTrue(result.isSuccessful());
+
+        CompilationUnit cu = result.getResult().get();
+        ClassOrInterfaceDeclaration implicitClass = cu.getClassByName("").get();
+        assertTrue(implicitClass.isCompact());
+        assertEquals(2, implicitClass.getMethods().size());
+    }
+
+    @Test
+    void publicMainMethod() {
+        String code = "public void main() { System.out.println(\"Public main\"); }";
+        ParseResult<CompilationUnit> result = javaParser.parse(COMPILATION_UNIT, provider(code));
+
+        assertNoProblems(result);
+        assertTrue(result.isSuccessful());
+
+        CompilationUnit cu = result.getResult().get();
+        ClassOrInterfaceDeclaration implicitClass = cu.getClassByName("").get();
+        assertTrue(implicitClass.isCompact());
+
+        MethodDeclaration mainMethod = implicitClass.getMethods().get(0);
+        assertTrue(mainMethod.isPublic());
+    }
+
+    @Test
+    void regularClassStillWorks() {
+        String code = "class HelloWorld { public static void main(String[] args) { } }";
+        ParseResult<CompilationUnit> result = javaParser.parse(COMPILATION_UNIT, provider(code));
+
+        assertNoProblems(result);
+        assertTrue(result.isSuccessful());
+
+        CompilationUnit cu = result.getResult().get();
+        ClassOrInterfaceDeclaration regularClass =
+                cu.getClassByName("HelloWorld").get();
+        assertFalse(regularClass.isCompact(), "Regular class should not be compact");
+    }
+
+    @Test
+    void mixedCompactAndRegularClass() {
+        String code = "void main() { new Helper().help(); }\n" + "class Helper { void help() { } }";
+        ParseResult<CompilationUnit> result = javaParser.parse(COMPILATION_UNIT, provider(code));
+
+        assertNoProblems(result);
+        assertTrue(result.isSuccessful());
+
+        CompilationUnit cu = result.getResult().get();
+        assertEquals(2, cu.getTypes().size(), "Should have compact class and regular class");
```

Comment:
`Helper` should be a nested class in the compact class. See my comment on the `java.jj` changes for a more detailed explanation.